src/prof.c

```diff
@@ -131,7 +131,7 @@ prof_alloc_rollback(tsd_t *tsd, prof_tctx_t *tctx, bool updated) {
 
 	cassert(config_prof);
 
-	if (updated) {
+	if (updated && tsd_reentrancy_level_get(tsd) == 0) {
```

Comment:
Can we reach here during reentrancy?

Reply:
I tried adding `assert(tsd_reentrancy_level_get(tsd) == 0);` and nothing broke.

I think logically `prof_alloc_rollback()` shouldn't be called by interrupting malloc, but our implementation doesn't forbid it.

First, my feeling is that the logic here for obtaining `tdata` and resetting the sampling threshold can be put under the `(uintptr_t)tctx > (uintptr_t)1U` branch below it. Thoughts?

If we keep the current logic unchanged, then by design the `tdata` creation and threshold reset part needs to be robust to the `(uintptr_t)tctx == (uintptr_t)1U` case, i.e. the no sample case, which is what's gonna happen in interrupting malloc calls. In `jemalloc.c`, `prof_alloc_rollback()` is in fact called whenever the internal allocation fails, regardless of whether there has been a sample or not; we'd probably never run into such an allocation failure case in the tests, but it might be possible in reality.

Reply:
Actually, I don't understand why we need to reset the sampling threshold here. If a sample was recorded, then the threshold should have already been reset; if a sample was not recorded, then we should continue counting rather than doing a reset which effectively serves to discard the accumulation so far. If there have been any reentrancy events, then we should count the allocations there rather than discarding them. So it seems to me that not having a reset here makes more sense in any case.

Reply:
I think the resetting is incorrect as well; let's tackle that later in another PR.

If we simply return from this function when reentrancy > 0, will it break anything? (like missing the destroy below?)